File: BSc/Data Structures and Algorithms II/MiniTrabalhos/P3/stack.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "stack.h"

struct stack *stack_new(int size)
{
    struct stack *stack = malloc(sizeof(struct stack));
    
    stack->pos = -1;
    stack->size = size;
    stack->array = (int*)malloc(stack->size * sizeof(int));

    return stack;

}

int isEmpty(struct stack *stack)
{
    if (stack->pos == -1)
    {
        return 1;
    }
    return -1;
}

int isFull(struct stack *stack)
{
    if (stack->pos == stack->size-1)
    {
        return 1;
    }
    return -1;
}

void push(struct stack *stack, int valor)
{
    if(isFull(stack)==-1)
    {
        stack->array[++stack->pos] = valor;
    }
}

int top (struct stack *stack)
{
    if (isEmpty(stack) == 1)
    {
        return -1;
    }
    return stack->array[stack->pos];
}

int pop(struct stack *stack)
{
    if (isEmpty(stack) == 1)
    {
        return -1;
    }
    return stack->array[stack->pos--];
}

File: BSc/Data Structures and Algorithms I/Implementacoes/Stack/Stack.java
package Stack;

public class Stack <E> implements iStack <E>{

    private E[] stack;
    int pos = 0;

    public Stack (int size)
    {
        this.stack = (E[])new Object[size];
    }

    @Override
    public void push(E o)
    {
        if (pos<stack.length)
        {
            stack[pos] = o;
            pos++;
        }
    }

    @Override
    public E top(){
        return stack[pos-1];
    }

    @Override
    public E pop() {
        pos--;
        return stack[pos];
    }

    @Override
    public int size() {
        return pos;
    }

    @Override
    public boolean empty()
    {
        if (pos!=0)
        {
            return false;
        }
        return true;
    }

    public String printStack()
    {
        String print= "";

        for(int i = 0; i<pos;i++)
        {
            print = print + stack [i] + "\n";
        }
        return print;
    }
}
